lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+
+namespace imex {
+
+mlir::StringRef getAllocSharedAttrName() { return "gpu.alloc_shared"; }
+
+struct InsertGPUAllocs
+    : public mlir::PassWrapper<InsertGPUAllocs,
+                               mlir::OperationPass<mlir::func::FuncOp>> {
+
+  InsertGPUAllocs() = default;
+  InsertGPUAllocs(bool gpuDealloc) { useGpuDealloc = gpuDealloc; }
+  InsertGPUAllocs(const InsertGPUAllocs &pass) : PassWrapper(pass) {}
+
+  virtual void
+  getDependentDialects(mlir::DialectRegistry &registry) const override {
+    registry.insert<mlir::memref::MemRefDialect>();
+    registry.insert<mlir::gpu::GPUDialect>();
+    registry.insert<mlir::scf::SCFDialect>();
+    registry.insert<mlir::arith::ArithmeticDialect>();
+  }
+
+  void runOnOperation() override {
+    auto func = getOperation();
+    auto &funcBody = func.getBody();
+    if (funcBody.empty()) {
+      return;
+    } else if (!llvm::hasSingleElement(funcBody)) {
+      func.emitError("Function must have exactly one block");
+      signalPassFailure();
+      return;
+    }
+
+    struct AccessType {
+      bool hostRead = false;
+      bool hostWrite = false;
+      bool deviceRead = false;
+      bool deviceWrite = false;
+    };
+
+    llvm::SmallMapVector<mlir::Operation *, AccessType, 8> gpuBufferAllocs;
+    llvm::SmallMapVector<unsigned, AccessType, 8> gpuBufferParams;
+    llvm::SmallMapVector<mlir::Operation *, AccessType, 8>
+        gpuGetMemrefGlobalParams;
+    auto &aliases = getAnalysis<mlir::BufferViewFlowAnalysis>();
+
+    auto getMemref = [](mlir::Operation *op)
+        -> llvm::Optional<mlir::SmallVector<mlir::Value, 4>> {
+      if (auto load = mlir::dyn_cast<mlir::memref::LoadOp>(op)) {
+        return {{load.memref()}};
+      } else if (auto store = mlir::dyn_cast<mlir::memref::StoreOp>(op)) {
+        return {{store.memref()}};
+      } else if (auto call = mlir::dyn_cast<mlir::func::CallOp>(op)) {
+        mlir::SmallVector<mlir::Value, 4> ret;
+        for (auto arg : call.operands()) {
+          if (arg.getType().isa<mlir::MemRefType>())
+            ret.emplace_back(arg);
+        }
+        return std::move(ret);
+      } else {
+        op->emitError("Uhhandled mem op in gpu region");
+        return llvm::None;
+      }
+    };
+
+    auto scfDialect = getContext().getOrLoadDialect<mlir::scf::SCFDialect>();
+
+    auto hasMemAccess = [](mlir::Operation *op) -> bool {
+      if (auto memInterface =
+              mlir::dyn_cast<mlir::MemoryEffectOpInterface>(op)) {
+        if (memInterface.hasEffect<mlir::MemoryEffects::Read>() ||
+            memInterface.hasEffect<mlir::MemoryEffects::Write>())
+          return true;
+      }
+      if (auto call = mlir::dyn_cast<mlir::func::CallOp>(op)) {
+        for (auto arg : call.operands()) {
+          if (arg.getType().isa<mlir::MemRefType>())
+            return true;
+        }
+      }
+      return false;
+    };
+
+    // Traverse through all the memory access ops under GPU launch Op
```

Comment:
Since you asked: until here, there is no comment or anything saying what happens in the code.

Reply:
Agree. Some reasonable comments are needed. Usually the questions from readers/reviewers are good hints where a comment might be needed.

Reply:
added more comments in the code